mev-boost-rs/src/relay.rs

```diff
@@ -0,0 +1,144 @@
+use beacon_api_client::Client as BeaconClient;
+use ethereum_consensus::{crypto::Error as CryptoError, primitives::BlsPublicKey};
+use thiserror::Error;
+use url::Url;
+
+use mev_rs::blinded_block_provider::Client;
+
+#[derive(Clone, Debug)]
+pub struct RelayEndpoint {
+    url: Url,
+    public_key: BlsPublicKey,
+}
+
+#[derive(Debug, Error)]
+pub enum RelayUrlError {
+    #[error("{0}")]
+    Bls(#[from] CryptoError),
+    #[error("{0}")]
+    Hex(#[from] hex::FromHexError),
+    #[error("public key {0} missing '0x' hex prefix")]
+    Missing0xPrefix(String),
+    #[error("URL {0} missing public key username")]
+    MissingPublicKey(String),
+}
+
+impl TryFrom<Url> for RelayEndpoint {
+    type Error = RelayUrlError;
+
+    fn try_from(url: Url) -> Result<Self, Self::Error> {
+        let public_key = url.username();
+        if public_key.is_empty() {
+            return Err(Self::Error::MissingPublicKey(url.to_string()));
+        }
+
+        let public_key =
+            public_key.strip_prefix("0x").ok_or(Self::Error::Missing0xPrefix(public_key.into()))?;
+        let public_key = hex::decode(public_key)?;
+        let public_key = BlsPublicKey::try_from(public_key.as_slice())?;
+
+        Ok(Self { url, public_key })
+    }
+}
+
+#[derive(Clone)]
+pub struct Relay {
+    api: Client,
```

Comment:
slight preference to just impl `Deref` for `Relay` that always yields `api` as a ref

then we can skip all of these `.api()` situations